ndpyramid/utils.py

```diff
@@ -29,3 +51,107 @@ def multiscales_template(
             'metadata': {'method': method, 'version': version, 'args': args, 'kwargs': kwargs},
         }
     ]
+
+
+def set_zarr_encoding(
+    ds: xr.Dataset,
+    codec_config: dict | None = None,
+    float_dtype: npt.DTypeLike | None = None,
+    int_dtype: npt.DTypeLike | None = None,
+) -> xr.Dataset:
+    """Set zarr encoding for each variable in the dataset
+
+    Parameters
+    ----------
+    ds : xr.Dataset
+        Input dataset
+    codec_config : dict, optional
+        Dictionary of parameters to pass to numcodecs.get_codec.
+        The default is {'id': 'zlib', 'level': 1}
+    float_dtype : str or dtype, optional
+        Dtype to cast floating point variables to
+
+    Returns
+    -------
+    ds : xr.Dataset
+        Output dataset with updated variable encodings
+    """
+    import numcodecs
+
+    ds = ds.copy()
+
+    if codec_config is None:
+        codec_config = {'id': 'zlib', 'level': 1}
+    compressor = numcodecs.get_codec(codec_config)
+
+    for k, da in ds.variables.items():
+
+        # maybe cast float type
+        if np.issubdtype(da.dtype, np.floating) and float_dtype is not None:
+            da = da.astype(float_dtype)
+
+        if np.issubdtype(da.dtype, np.integer) and int_dtype is not None:
+            da = da.astype(int_dtype)
+
+        # remove old encoding
+        da.encoding.clear()
+
+        # update with new encoding
+        da.encoding['compressor'] = compressor
+        with contextlib.suppress(KeyError):
+            del da.attrs['_FillValue']
+        da.encoding['_FillValue'] = default_fillvals.get(
+            da.dtype.str[-2:], None
+        )  # TODO: handle date/time types
+
+        ds[k] = da
+
+    return ds
+
+
+def _add_metadata_and_zarr_encoding(
```

Comment:
I think we can make this a public utility function

```suggestion
def add_metadata_and_zarr_encoding(
```